### ISSUE TITLE
Memory leak due to `Bsonc::new` behavior

### ISSUE THREAD

**Issue by austinjones** [appsignal/mongo-rust-driver#66]
Hi,
I found a memory leak in `mongo_driver`, and narrowed it down to the behavior of `Bsonc::new`.  Several commands call `Bsonc::new` to create a 'reply' document so that the C driver has a bson target to write into.  A good example is: https://github.com/appsignal/mongo-rust-driver/blob/main/src/collection.rs#L324

I think that the document is never dropped, and if something is read (e.g. calling command_simple with 'find'), then the leak can be large.  Here is a simple example where the leak is extreme (100s of MB): https://github.com/austinjones/mongo-driver-leak/blob/master/src/main.rs

I have [a fix](https://github.com/austinjones/mongo-rust-driver/commit/13a19f983048979cf9b2fc5a28982194a1c4a2e1) for this issue.  Let me know if you'd like a PR!

**Comment by thijsc** [comment]
I did just make some changes in this area. The newer version of the C driver takes ownership of the bsonc in some cases. So bsoncs were double-freed. I then made the drop logic optional for those cases.

It's a bit hard to see what your fix does since it formats a lot of code differently. What is the change that fixes it?

**Comment by austinjones** [comment]
The only functional change was replacing the `Bsonc::new()` implementation.  The `bson_t` construction is the same, but `destroy_inner_on_drop` is now true.  Here is the [line](https://github.com/austinjones/mongo-rust-driver/commit/13a19f983048979cf9b2fc5a28982194a1c4a2e1#diff-5aacc217a933bca06205eeaa1c2a1c2eafad289dd5dba1b518ec1cbf3641b165R25).
```
    pub fn new() -> Bsonc {
        Bsonc::from_ptr(unsafe { bindings::bson_new() })
        let inner: *const bindings::bson_t = unsafe { bindings::bson_new() };
        assert!(!inner.is_null());

        Bsonc {
            inner: inner as *mut bindings::bson_t,
            destroy_inner_on_drop: true,
        }
    }
```

Also, direct calls to `from_ptr` seem fine (e.g. Cursor iterator) - no leaks there.

**Comment by thijsc** [comment]
I implemented your fix in #67. Will release that once we get a green build.

**Comment by thijsc** [comment]
A fix has been released: 7fcdeed53ce03b00d228df804b09be1c360a1ce8

Thanks for figuring this out!
